### DISCUSSION TITLE
Use hit object count for classic score instead of max combo

### DISCUSSION THREAD

**Discussion by respektive** [ppy/osu#16931] [Ideas] [answered]
As classic score is trying to emulate scorev1 using hit objects instead of max combo in the calculation makes more sense and gives more similar results.  

Using max combo results in some maps being fairly broken, like slider heavy maps or lower difficulties (because of the amount of sliderticks).  
This can result in over 100 million score difference in extreme cases like on this map.  

[Ocelot - KAEDE [EX EX]](https://osu.ppy.sh/beatmapsets/660630#osu/1398809)

| player | stable | master | my proposal |
| --- | --- | --- | --- |
| [Woey](https://osu.ppy.sh/scores/osu/3364956481) | 29,181,048 | 130,675,126 | 7,117,802 |
> Note that KAEDE is an edge case cause of heavy repeat slider use but still is much closer to the scorev1 value than before.

So my idea is changing the max combo part in the formula to use the amount of hit objects instead like so:

![formulas](https://user-images.githubusercontent.com/22381687/154852054-8bbfd46d-e7e0-46e1-b4c9-d24f154366db.png)

I've done some testing using this new formula and found the values to be much closer to scorev1. 

[xi - Blue Zenith [Akaphyxia's Normal]](https://osu.ppy.sh/beatmapsets/292301#osu/688710)

| player | stable | master | my proposal |
| --- | --- | --- | --- |
| [Rachild](https://osu.ppy.sh/scores/osu/3686967072) | 8,881,830 | 21,793,886 | 12,112,073 |
| [ieu-](https://osu.ppy.sh/scores/osu/3582494383) | 8,481,769 | 19,590,299 | 10,887,418 |

[Ne Obliviscaris - Painters of the Tempest (Part II)](https://osu.ppy.sh/beatmapsets/724192#osu/1529189)

| player | stable | master | my proposal |
| --- | --- | --- | --- |
| [ASecretBox](https://osu.ppy.sh/scores/osu/3214803949) | 1,654,213,675 | 1,399,513,766 | 1,627,218,635 |
| [Aricin](https://osu.ppy.sh/scores/osu/2832521487) | 1,571,805,970 | 1,251,469,202 | 1,455,086,799 |

Here is some screenshots showing the difference in the overall score ranking showing a clear difference:
![score ranking](https://user-images.githubusercontent.com/22381687/154853662-29a09b81-f0bc-47fe-9bc6-bcbf30ce8d47.png)
> this was calculated using the dumps released by peppy and self gathered data for players not in dumps

**Comment by WitherFlower** [comment]
For more precision regarding how the `*36` coefficient was chosen, I ran a calculation on a diverse set of maps (different lengths and star ratings) and used the sum of the squared relative errors to find the best coefficient possible.

![image](https://user-images.githubusercontent.com/52672543/154854714-5275b45e-9e26-42df-a52a-b16fef624a0f.png)

This also shows that the error is inferior using hitobjects instead of combo in the formula.

The corresponding spreadsheet can be found here : https://docs.google.com/spreadsheets/d/1hYsT3U3b0tg9SIMDhmBmqGBVTXASufQ8W0sQ-toifTg/edit#gid=0

**Comment by respektive** [comment]
I forgot to mention that this would also resolve this issue https://github.com/ppy/osu/issues/14697

**Comment by smoogipoo** [comment]
Sounds good, but I think this is going to require some work, combo's easy because we always have that available in some form either from the API or by running diffcalc locally. The number of hitobjects isn't standardised across all difficulty calculators and is not currently available via the API I believe.
Arguably the API should be the one providing us total scores for beatmap listing leaderboards.

**Reply by peppy** [reply]
We do have hit counts available in the database (circle/slider/spinner). Is the issue that this differs between rulesets?

**Reply by smoogipoo** [reply]
Indeed they are available... I had a brain fart.

**Comment by smoogipoo** [comment, marked as answer]
I've implemented this in https://github.com/ppy/osu/pull/16938.
